File: scripts/generate-manifest.mjs
import { promises as fs } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const projectRoot = path.resolve(__dirname, "..");
const scanDirs = ["about", "data", "home", "object", "photo", "root"];

const manifest = {
  generatedAt: new Date().toISOString(),
  files: [],
};

for (const dir of scanDirs) {
  const absoluteDir = path.join(projectRoot, dir);
  const exists = await pathExists(absoluteDir);
  if (!exists) {
    continue;
  }
  const files = await walkDirectory(absoluteDir);
  files.forEach((filePath) => {
    const relative = toPosix(path.relative(projectRoot, filePath));
    const name = path.basename(filePath);
    manifest.files.push({
      path: relative,
      dir,
      name,
      baseName: path.basename(filePath, path.extname(filePath)),
      extension: path.extname(filePath).toLowerCase(),
    });
  });
}

manifest.files.sort((a, b) => a.path.localeCompare(b.path));

const outputPath = path.join(projectRoot, "root", "site-manifest.json");
await fs.mkdir(path.dirname(outputPath), { recursive: true });
await fs.writeFile(outputPath, JSON.stringify(manifest, null, 2) + "\n", "utf8");
console.log(`Manifest written to ${outputPath}`);

async function walkDirectory(dir) {
  const entries = await fs.readdir(dir, { withFileTypes: true });
  const files = [];

  for (const entry of entries) {
    if (entry.name.startsWith(".")) {
      continue;
    }
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      files.push(...(await walkDirectory(fullPath)));
    } else if (entry.isFile()) {
      files.push(fullPath);
    }
  }

  return files;
}

async function pathExists(targetPath) {
  try {
    await fs.access(targetPath);
    return true;
  } catch {
    return false;
  }
}

function toPosix(value) {
  return value.split(path.sep).join("/");
}

File: scripts/generate_manifest.py
from __future__ import annotations

import json
from datetime import datetime, timezone
from pathlib import Path


PROJECT_ROOT = Path(__file__).resolve().parent.parent
SCAN_DIRS = ["about", "data", "home", "object", "photo", "root"]
OUTPUT_FILE = PROJECT_ROOT / "root" / "site-manifest.json"


def main() -> None:
    manifest = {
        "generatedAt": datetime.now(timezone.utc).isoformat(),
        "files": [],
    }

    for directory_name in SCAN_DIRS:
        directory = PROJECT_ROOT / directory_name
        if not directory.exists():
            continue

        for file_path in sorted(path for path in directory.rglob("*") if path.is_file()):
            if any(part.startswith(".") for part in file_path.parts):
                continue

            manifest["files"].append(
                {
                    "path": file_path.relative_to(PROJECT_ROOT).as_posix(),
                    "dir": directory_name,
                    "name": file_path.name,
                    "baseName": file_path.stem,
                    "extension": file_path.suffix.lower(),
                }
            )

    OUTPUT_FILE.parent.mkdir(parents=True, exist_ok=True)
    OUTPUT_FILE.write_text(
        json.dumps(manifest, ensure_ascii=False, indent=2) + "\n",
        encoding="utf-8",
    )
    print(f"Manifest written to {OUTPUT_FILE}")


if __name__ == "__main__":
    main()
